source/fuzz/fuzzer.cpp

```diff
@@ -165,200 +268,46 @@ bool Fuzzer::ApplyPassAndCheckValidity(FuzzerPass* pass) const {
-    MaybeAddRepeatedPass<FuzzerPassOutlineFunctions>(&pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassPermuteBlocks>(&pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassPermuteFunctionParameters>(&pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassPermuteInstructions>(&pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassPropagateInstructionsDown>(&pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassPropagateInstructionsUp>(&pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassPushIdsThroughVariables>(&pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassReplaceAddsSubsMulsWithCarryingExtended>(
-        &pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassReplaceBranchesFromDeadBlocksWithExits>(
-        &pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassReplaceCopyMemoriesWithLoadsStores>(
-        &pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassReplaceCopyObjectsWithStoresLoads>(
-        &pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassReplaceLoadsStoresWithCopyMemories>(
-        &pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassReplaceParameterWithGlobal>(&pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassReplaceLinearAlgebraInstructions>(
-        &pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassReplaceOpPhiIdsFromDeadPredecessors>(
-        &pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassReplaceOpSelectsWithConditionalBranches>(
-        &pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassReplaceParamsWithStruct>(&pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassSplitBlocks>(&pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassSwapBranchConditionalOperands>(
-        &pass_instances);
-    MaybeAddRepeatedPass<FuzzerPassWrapRegionsInSelections>(&pass_instances);
-    // There is a theoretical possibility that no pass instances were created
-    // until now; loop again if so.
-  } while (pass_instances.GetPasses().empty());
-
-  RepeatedPassRecommenderStandard pass_recommender(&pass_instances,
-                                                   fuzzer_context_.get());
-
-  std::unique_ptr<RepeatedPassManager> repeated_pass_manager = nullptr;
-  switch (repeated_pass_strategy_) {
-    case RepeatedPassStrategy::kSimple:
-      repeated_pass_manager = MakeUnique<RepeatedPassManagerSimple>(
-          fuzzer_context_.get(), &pass_instances);
-      break;
-    case RepeatedPassStrategy::kLoopedWithRecommendations:
-      repeated_pass_manager =
-          MakeUnique<RepeatedPassManagerLoopedWithRecommendations>(
-              fuzzer_context_.get(), &pass_instances, &pass_recommender);
-      break;
-    case RepeatedPassStrategy::kRandomWithRecommendations:
-      repeated_pass_manager =
-          MakeUnique<RepeatedPassManagerRandomWithRecommendations>(
-              fuzzer_context_.get(), &pass_instances, &pass_recommender);
-      break;
+  if (num_repeated_passes_applied_ >=
+      fuzzer_context_->GetTransformationLimit()) {
+    return {FuzzerResultStatus::kFuzzerStuck, {}};
```

Comment:
How about we move this into the do ... while loop, and then remove the similar check from "ShouldContinueFuzzing"? That would mean that if the fuzzer really does get stuck (because fuzzer passes are not able to do anything) we will always return kFuzzerStuck? Right now we'll only return kFuzzerStuck if we find we're stuck at the start of Run.

We don't have to necessarily regard the fuzzer getting stuck as a bug, but it's nice to be able to know when it's happened.

Reply:
Good idea. We can also do the same thing when we check if the number of transformations reached its limit. What do you think?

Reply:
Yes, i.e. we'd move this code out of the method and to the start of the loop (and remove the related code from before the loop):

```
  if (num_of_transformations_to_apply != 0 &&
      transformations_applied_so_far - initial_num_of_transformations >=
          num_of_transformations_to_apply) {
    // Stop because we've applied the maximum number of transformations for a
    // single execution of a |Run| method.
    return false;
  }
```

and we would make it instead return kTransformationLimitReached.

I wonder whether we should also move this?

```
  if (num_of_transformations_to_apply != 0 &&
      transformations_applied_so_far - initial_num_of_transformations >=
          num_of_transformations_to_apply) {
    // Stop because we've applied the maximum number of transformations for a
    // single execution of a |Run| method.
    return false;
  }
```

Then the loop would have the form:

```
do {
  if (FuzzerIsStuck(...)) {
    return kFuzzerStuck;
  }
  if (TransformationLimitReached(...)) { // This would capture reaching the global limit, as well as the run-specific limit
    return {kTransformationLimitReached, {}}
  }
  if (!ApplyPassAndCheckValidity(...)) {
    return {kFuzzerPassLedToInvalidModule, {}};
  }
} while (ShouldContinueFuzzing(...));
```

And maybe we rename `ShouldContinueFuzzing` to `ShouldContinueRepeatedPasses`.

Reply:
Fixed in 1f68fec.